src/ZoneServer/Skills/Handlers/Sapper/DetonateTraps.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Threading.Tasks;
+using Melia.Shared.L10N;
+using Melia.Shared.Tos.Const;
+using Melia.Shared.World;
+using Melia.Zone.Network;
+using Melia.Zone.Skills.Handlers.Base;
+using Melia.Zone.Skills.SplashAreas;
+using Melia.Zone.World.Actors;
+using Melia.Zone.World.Actors.CombatEntities.Components;
+using Melia.Zone.Skills.Combat;
+using Melia.Zone.World.Actors.Characters;
+using Melia.Zone.World.Actors.Monsters;
+using static Melia.Zone.Skills.SkillUseFunctions;
+using Melia.Shared.Data.Database;
+using System.Threading;
+using System.Collections.Generic;
+
+namespace Melia.Zone.Skills.Handlers.Sapper
+{
+	/// <summary>
+	/// Handler for the Sapper skill Detonate Traps.
+	/// </summary>
+	[SkillHandler(SkillId.Sapper_DetonateTraps)]
+	public class DetonateTraps : IGroundSkillHandler
+	{
+		public void Handle(Skill skill, ICombatEntity caster, Position originPos, Position farPos, ICombatEntity target)
+		{
+			if (!caster.TrySpendSp(skill))
+			{
+				caster.ServerMessage(Localization.Get("Not enough SP."));
+				return;
+			}
+
+			if (!caster.Position.InRange2D(farPos, skill.Data.MaxRange))
+			{
+				caster.ServerMessage(Localization.Get("Too far away."));
+				return;
+			}
+
+			skill.IncreaseOverheat();
+			caster.SetAttackState(true);
+
+			Send.ZC_SKILL_READY(caster, skill, caster.Position, caster.Position);
+			Send.ZC_NORMAL.UpdateSkillEffect(caster, caster.Handle, caster.Position, caster.Direction, caster.Position);
+			Send.ZC_SKILL_MELEE_GROUND(caster, skill, caster.Position, ForceId.GetNew(), null);
+
+			var character = caster as Character;
+			var removeList = new List<Mob>();
+
+			foreach (var trap in character.PlacedTraps)
+			{
+				Send.ZC_DEAD(trap, trap.Position);
+				caster.Map.RemoveMonster(trap);
+				if (character.PlacedTraps.Contains(trap))
```

Comment:
Isn't this check pointless, since you're iterating over the traps? It has to be in the list.